src/pg_copy.c

```diff
@@ -0,0 +1,742 @@
+	 * converts it.	 Note: we guarantee that there is a \0 at
+	 * raw_buf[raw_buf_len].
+	 */
+#define RAW_BUF_SIZE 65536      /* we palloc RAW_BUF_SIZE+1 bytes */
+	char *raw_buf;
+	int raw_buf_index;          /* next byte to process */
+	int raw_buf_len;            /* total # of bytes stored */
+} CopyStateData;
+
+/* TODOL should be moved to copy.c */
+static StringInfo
+CopyGetLineBuf(CopyStateData *cs)
+{
+	return &cs->line_buf;
+}
+
+
+#define MAX_SHARDS 1001
+
+static uint32
+shard_id_hash_fn(const void *key, Size keysize)
+{
+	return *(int32 *) key ^ *((int32 *) key + 1);
+}
+
+
+/*
+ * Construct hash table used for shardId->Connection mapping.
+ * We can not use connection cache from connection.c used by GteConnection because
+ * we need to establish multiple connections with each nodes: one connection per shard
+ */
+static HTAB *
+CreateShardToConnectionHash()
+{
+	HASHCTL info;
+
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(ShardId);
+	info.entrysize = sizeof(ShardConnections);
+	info.hash = shard_id_hash_fn;
+
+	return hash_create("shardToConn", MAX_SHARDS, &info, HASH_ELEM | HASH_FUNCTION);
+}
+
+
+static char const *
+ConstructCopyStatement(CopyStmt *copyStatement, ShardId shardId)
+{
+	StringInfo buf = makeStringInfo();
+	ListCell *cell = NULL;
+	char sep = '\0';
+	char const *qualifiedName = quote_qualified_identifier(
+		copyStatement->relation->schemaname,
+		copyStatement->relation->relname);
+	appendStringInfo(buf, "COPY %s_%ld ", qualifiedName, (long) shardId);
+	if (copyStatement->attlist)
+	{
+		sep = '(';
+		foreach(cell, copyStatement->attlist)
+		{
+			appendStringInfoChar(buf, sep);
+			appendStringInfoString(buf, strVal(lfirst(cell)));
+			sep = ',';
+		}
+		appendStringInfoChar(buf, ')');
+	}
+	appendStringInfoString(buf, "FROM STDIN");
+
+	if (copyStatement->options)
+	{
+		appendStringInfoString(buf, " WITH ");
+		sep = '(';
+		foreach(cell, copyStatement->options)
+		{
+			DefElem *def = (DefElem *) lfirst(cell);
+			appendStringInfo(buf, "%c%s '%s'", sep, def->defname, defGetString(def));
+			sep = ',';
+		}
+		appendStringInfoChar(buf, ')');
+	}
+	return buf->data;
+}
+
+static bool
+PgCopyEnd(PGconn *con, char const* msg)
+{
+	PGresult *result = NULL;
+	if (PQputCopyEnd(con, msg) == 1)
+	{
+		while ((result = PQgetResult(con)) != NULL)
+		{
+			if (PQresultStatus(result) != PGRES_COMMAND_OK)
+			{
+				return false;
+			}
+			PQclear(result);
+		}
+		return true;
+	}
+	return false;
+}
+
+static bool
+PgCopyPrepareTransaction(ShardId shardId, PGconn *conn, void *arg, bool status)
+{
+	PgShardTransactionManager const *tmgr =
+		&PgShardTransManagerImpl[PgShardCurrTransManager];
+	return PgCopyEnd(conn, NULL)
+		&& tmgr->Prepare(conn);
+}
+
+
+static bool
+PgCopyAbortTransaction(ShardId shardId, PGconn *conn, void *arg, bool status)
+{
+	PgShardTransactionManager const *tmgr =
+		&PgShardTransManagerImpl[PgShardCurrTransManager];
+
+	if (status)
```

Comment:
I still don't like the name status, should have a semantic name like sentPrepare.